packages/rest/src/parser.ts

```diff
@@ -75,32 +89,112 @@ export async function parseOperationArgs(
+    value: undefined,
+  };
+  if (!operationSpec.requestBody) return Promise.resolve(requestBody);
 
-  const contentType = getContentType(request);
+  debug('Request body parser options: %j', options);
+
+  const contentType = getContentType(request) || 'application/json';
   debug('Loading request body with content type %j', contentType);
-  if (contentType && !/json/.test(contentType)) {
+
+  // the type of `operationSpec.requestBody` could be `RequestBodyObject`
+  // or `ReferenceObject`, resolving a `$ref` value is not supported yet.
+  if (isReferenceObject(operationSpec.requestBody)) {
+    throw new Error('$ref requestBody is not supported yet.');
+  }
+
+  let content = operationSpec.requestBody.content || {};
+  if (!Object.keys(content).length) {
+    content = {
+      // default to allow json and urlencoded
+      'application/json': {schema: {type: 'object'}},
+      'application/x-www-form-urlencoded': {schema: {type: 'object'}},
+    };
+  }
+
+  // Check of the request content type matches one of the expected media
+  // types in the request body spec
+  let matchedMediaType: string | false = false;
+  for (const type in content) {
+    matchedMediaType = is(contentType, type);
+    if (matchedMediaType) {
+      requestBody.mediaType = type;
+      requestBody.schema = content[type].schema;
+      break;
+    }
+  }
+
+  if (!matchedMediaType) {
+    // No matching media type found, fail fast
     throw new HttpErrors.UnsupportedMediaType(
-      `Content-type ${contentType} is not supported.`,
+      `Content-type ${contentType} does not match [${Object.keys(content)}].`,
     );
   }
 
-  return await parseJsonBody(request).catch((err: HttpError) => {
-    debug('Cannot parse request body %j', err);
-    err.statusCode = 400;
-    throw err;
-  });
+  if (is(matchedMediaType, 'urlencoded')) {
+    try {
+      const body = await parseFormBody(
+        request,
+        // use `qs` modules to handle complex objects
+        Object.assign(
+          {
+            querystring: {
+              parse: (str: string, cb: Function) => {
+                cb(null, qs.parse(str));
+              },
+            },
+          },
+          options,
+        ),
+      );
+      return Object.assign(requestBody, {
+        // form parser returns an object without prototype
+        // create a new copy to simplify shouldjs assertions
+        value: Object.assign({}, body),
+        // urlencoded body only provide string values
+        // set the flag so that AJV can coerce them based on the schema
+        coercionRequired: true,
+      });
+    } catch (err) {
+      throw normalizeParsingError(err);
+    }
+  }
+
+  if (is(matchedMediaType, 'json')) {
+    try {
+      const jsonBody = await parseJsonBody(request, options);
+      requestBody.value = jsonBody;
+      return requestBody;
+    } catch (err) {
+      throw normalizeParsingError(err);
+    }
+  }
+
+  throw new HttpErrors.UnsupportedMediaType(
```

Comment:
There is a file to organize the HttpErrors we defined: https://github.com/strongloop/loopback-next/blob/a478e567daaf732098ca2c1b9994a479146208fe/packages/rest/src/rest-http-error.ts
Defining errors there makes error messages re-usable, e.g. assert the error and message in test.